File: core/src/commonMain/kotlin/work/socialhub/kslack/api/methods/request/views/ViewsOpenRequest.java
package com.github.seratch.jslack.api.methods.request.views;

import com.github.seratch.jslack.api.methods.SlackApiRequest;
import com.github.seratch.jslack.api.model.view.View;

public class ViewsOpenRequest implements SlackApiRequest {
    private String token;
    private String triggerId;
    private View view;
    private String viewAsString;

    ViewsOpenRequest(String token, String triggerId, View view, String viewAsString) {
        this.token = token;
        this.triggerId = triggerId;
        this.view = view;
        this.viewAsString = viewAsString;
    }

    public static ViewsOpenRequestBuilder builder() {
        return new ViewsOpenRequestBuilder();
    }

    public String getToken() {
        return this.token;
    }

    public String getTriggerId() {
        return this.triggerId;
    }

    public View getView() {
        return this.view;
    }

    public String getViewAsString() {
        return this.viewAsString;
    }

    public void setToken(String token) {
        this.token = token;
    }

    public void setTriggerId(String triggerId) {
        this.triggerId = triggerId;
    }

    public void setView(View view) {
        this.view = view;
    }

    public void setViewAsString(String viewAsString) {
        this.viewAsString = viewAsString;
    }

    public static class ViewsOpenRequestBuilder {
        private String token;
        private String triggerId;
        private View view;
        private String viewAsString;

        ViewsOpenRequestBuilder() {
        }

        public ViewsOpenRequest.ViewsOpenRequestBuilder token(String token) {
            this.token = token;
            return this;
        }

        public ViewsOpenRequest.ViewsOpenRequestBuilder triggerId(String triggerId) {
            this.triggerId = triggerId;
            return this;
        }

        public ViewsOpenRequest.ViewsOpenRequestBuilder view(View view) {
            this.view = view;
            return this;
        }

        public ViewsOpenRequest.ViewsOpenRequestBuilder viewAsString(String viewAsString) {
            this.viewAsString = viewAsString;
            return this;
        }

        public ViewsOpenRequest build() {
            return new ViewsOpenRequest(token, triggerId, view, viewAsString);
        }

        public String toString() {
            return "ViewsOpenRequest.ViewsOpenRequestBuilder(token=" + this.token + ", triggerId=" + this.triggerId + ", view=" + this.view + ", viewAsString=" + this.viewAsString + ")";
        }
    }
}

File: core/src/commonMain/kotlin/work/socialhub/kslack/api/methods/request/views/ViewsOpenRequest.kt
package work.socialhub.kslack.api.methods.request.views

import work.socialhub.kslack.api.methods.FormRequest
import work.socialhub.kslack.api.methods.SlackApiRequest
import work.socialhub.kslack.api.methods.helper.JsonHelper.toJson
import work.socialhub.kslack.entity.view.View
import kotlin.js.JsExport

@JsExport
class ViewsOpenRequest(
    override var token: String?,
    var triggerId: String?,
    var view: View?,
    var viewAsString: String?
) : SlackApiRequest, FormRequest {

    override fun toMap(): Map<String, Any> {
        return mutableMapOf<String, Any>().also {
            it.addParam("trigger_id", triggerId)
            if (viewAsString != null) {
                it.addParam("view", viewAsString)
            } else {
                it.addParam("view", toJson(view))
            }
        }
    }
}